stac_fastapi/api/stac_fastapi/api/models.py

```diff
@@ -123,6 +124,8 @@ class ItemCollectionUri(CollectionUri):
     """Get item collection."""
 
     limit: int = attr.ib(default=10)
+    bbox: Optional[str] = attr.ib(default=None, converter=str2list)
```

Comment:
Wouldn't it make sense to convert this directly to e.g. `Optional[List[float]]` for bbox and `Optional[List[Union[datetime, None]]]` for datetime so that you don't need to handle the conversion issues downstream?

Reply:
Yes I mentioned that in my comment above point 3.

I held off on changing it until I get some feedback, as that is the current behavior in the similar routes.

Reply:
`datetime` is a little more complicated, because it also supports ranges like `2020-01-01/2022-01-01` or `2020-01-01/..`, so it must allow `str` input (or a clever pydantic type enforcing either datetime or something with a slash in it).

Reply:
Ah, I did not see that comment. Still, a list of datetime/None should work. None could be used for open date ranges as specified also for STAC extents in collections...

Reply:
Oh sorry I misunderstood you. Yes that makes sense, but as I've looked into this I've found some inconsistency.

For `sqlalchemy`, I [added code](https://github.com/stac-utils/stac-fastapi/pull/477/commits/a68abdfd2085681df2923402f0fe779c7d21831c) based on the [post_search](https://github.com/carderne/stac-fastapi/blob/3ea9232a5c741cb0389410a18e0deed954be1203/stac_fastapi/sqlalchemy/stac_fastapi/sqlalchemy/core.py#L385-L399) implementation: it has some complicated logic on deciding how to treat the `datetime` parameter, and _treats a single datetime as an exact datetime_. It's also more lenient on datetime formatting.

For `pgstac`, I [just used](https://github.com/stac-utils/stac-fastapi/pull/477/commits/26cd5f3bebd08aa1f8f7bdb0c9c144848d2636eb) `post_request_model`, which [treats a single value as an end-date](https://github.com/carderne/stac-fastapi/blob/3ea9232a5c741cb0389410a18e0deed954be1203/stac_fastapi/types/stac_fastapi/types/search.py#L157-L158), and is stricter on enforcing RFC3339.

So a few questions:
1. Should we update the `sqlalchemy` `post_search` and `item_collection` routes to use `post_request_model`?
2. Should a single datetime be considered as an end date? It seems much more useful (if less obvious) than the unlikely case you're able to hit on the exact datetime of an item...
3. If the latter, datetime could be typed as (being pedantic): `None | tuple[datetime, datetime] | tuple[datetime, None] | tuple[None, datetime]`. This is all already validated in `post_request_model` (see the last link above). I think actually enforcing this typing would break the pgstac implementation, as it passes the raw `datetime/datetime` string to the DB's `search` function.

Reply:
> Should we update the sqlalchemy post_search and item_collection routes to use post_request_model?

I think the sqlalchemy backend would be better if, like pgstac, it had one function (`_search_base`) which contains most of the important query logic and is wrapped by all of the other methods.

> Should a single datetime be considered as an end date? It seems much more useful (if less obvious) than the unlikely case you're able to hit on the exact datetime of an item...

My interpretation of the [OAF datetime parameter](http://docs.opengeospatial.org/is/17-069r3/17-069r3.html#_parameter_datetime) is a single date time is an exact date time (`==`) while date time ranges are either `../end_date` or `start_date/..`.  I agree that `==` is arguably less useful when searching than a datetime range but the distinction is there to let users do both.

> If the latter, datetime could be typed as (being pedantic): None | tuple[datetime, datetime] | tuple[datetime, None] | tuple[None, datetime]. This is all already validated in post_request_model (see the last link above). I think actually enforcing this typing would break the pgstac implementation, as it passes the raw datetime/datetime string to the DB's search function.

Lol I should have read this thread first, I proposed the same thing in PR comment below.  We shouldn't be too concerned about breaking backend implementations w/ changes to the API layer as long as those changes make the API layer stronger and improve separation of concerns (I think here they do).  Backend implementations shouldn't have to guess what information is included in a request, that information should be presented in a way that is clear and easy to use and I'm not in love with the current solution for datetime.  Pgstac can always coerce these back to a string.